src/OpenDiffix.Core/Aggregator.fs

```diff
@@ -382,6 +382,125 @@ type private DiffixSumNoise(summandType, aidsCount) =
+        state.[other.Key] <- current + other.Value
+
+    member this.Final(aggContext, anonContext) =
+      let bins = Dictionary<int64, int64>()
+
+      for pair in state do
+        let binLabel = generalize pair.Value
+        bins |> Dictionary.increment binLabel
+
+      bins
+      |> Seq.map (fun pair -> (pair.Key, pair.Value))
+      |> Seq.sortBy fst
+      |> Seq.map (fun (binLabel, aidCount) -> Value.List [ Integer binLabel; Integer aidCount ])
+      |> Seq.toList
+      |> Value.List
+
+  member this.State = state
+
+
+type private CountHistogramAidTracker = { mutable RowCount: int64; LowCount: DiffixLowCount }
+
+type private CountHistogramBin = { mutable AidCount: int64; LowCount: DiffixLowCount }
+
+type private DiffixCountHistogram(aidsCount, binSize: int64 option) =
+  let state = Dictionary<AidHash, CountHistogramAidTracker>()
+
+  let makeAidTracker () =
+    { RowCount = 0; LowCount = DiffixLowCount(aidsCount) }
+
+  let generalize count =
+    match binSize with
+    | None
+    | Some 1L -> count
+    | Some binSize -> (float count / float binSize) |> floor |> int64 |> (*) binSize
+
+  interface IAggregator with
+    member this.Transition args =
+      match args with
+      | Value.List [] :: _ -> invalidArgs args
+      | _aidInstances :: Null :: _ -> () // Ignore NULLs
+      | aidInstances :: aid :: _ ->
+        let aidState = state |> Dictionary.getOrInit (hashAid aid) makeAidTracker
+        aidState.RowCount <- aidState.RowCount + 1L
+        (aidState.LowCount :> IAggregator).Transition([ aidInstances ])
+      | _ -> invalidArgs args
+
+    member this.Merge aggregator =
+      let otherState = (castAggregator<DiffixCountHistogram> aggregator).State
+
+      for other in otherState do
+        let currentState = state |> Dictionary.getOrInit other.Key makeAidTracker
+        currentState.RowCount <- currentState.RowCount + other.Value.RowCount
+        (currentState.LowCount :> IAggregator).Merge(other.Value.LowCount)
+
+    member this.Final(aggContext, anonContext) =
+      let bins = Dictionary<int64, CountHistogramBin>()
+
+      for pair in state do
+        let aidEntry = pair.Value
+        let binLabel = generalize aidEntry.RowCount
+
+        match bins.TryGetValue(binLabel) with
+        | true, bin ->
+          // Bin already exists, merge to it.
+          (bin.LowCount :> IAggregator).Merge(aidEntry.LowCount)
+          bin.AidCount <- bin.AidCount + 1L
+        | false, _ ->
+          // New bin, move ownership from current tracker because we don't need it any further.
+          bins.[binLabel] <- { AidCount = 1L; LowCount = aidEntry.LowCount }
+
+      let suppressBin = { AidCount = 0L; LowCount = DiffixLowCount(aidsCount) }
+
+      let highCountBins =
+        bins
+        |> Seq.choose (fun pair ->
+          let binLabel = pair.Key
+          let bin = pair.Value
+
+          if Anonymizer.isLowCount aggContext.AnonymizationParams bin.LowCount.State then
```

Comment:
do you think moving this bit to `Anonymizer.countHistogram` would be justified or overengineered. I feel like for consistency, some portion of this code should already be living there, even if it's going to be rather trivial at the moment.

Reply:
No thoughts on this. I would prefer to leave for later until we add the additional missing features.

Reply:
I moved only the noise adding part in `Anonymizer` the Dictionary handling remains there because I don't see how it can be reused.